File: api/internal/repository/memory/inmemory_activity_repository.go
package memory

import (
	"time"

	"github.com/idoceb00/elogap-api/internal/domain"
	"github.com/idoceb00/elogap-api/internal/repository"
)

type InMemoryActivityRepository struct {
	data []domain.Activity
}

func NewInMemoryActivityRepository() repository.ActivityRepository {
	now := time.Now().UTC()

	return &InMemoryActivityRepository{
		data: []domain.Activity{
			{
				ID:        "match_1",
				PlayerID:  "player_1",
				Result:    domain.ResultWin,
				QueueType: "RANKED_SOLO",
				Champion:  "Ahri",
				Role:      "MID",
				Kills:     10,
				Deaths:    2,
				Assists:   8,
				CS:        210,
				Duration:  1714,
				Damage:    24500,
				Vision:    32,
				PlayedAt:  now.Add(-2 * time.Hour),
			},
			{
				ID:        "match_2",
				PlayerID:  "player_1",
				Result:    domain.ResultLoss,
				QueueType: "RANKED_SOLO",
				Champion:  "Jinx",
				Role:      "ADC",
				Kills:     4,
				Deaths:    7,
				Assists:   6,
				CS:        260,
				Duration:  1980,
				Damage:    30200,
				Vision:    18,
				PlayedAt:  now.Add(-6 * time.Hour),
			},
		},
	}
}

func (r *InMemoryActivityRepository) List() ([]domain.Activity, error) {
	// return a copy to avoid accidental mutation
	out := make([]domain.Activity, len(r.data))
	copy(out, r.data)
	return out, nil
}

func (r *InMemoryActivityRepository) FindByID(id string) (*domain.Activity, error) {
	for i := range r.data {
		if r.data[i].ID == id {
			// return copy
			a := r.data[i]
			return &a, nil
		}
	}
	return nil, repository.ErrNotFound
}

File: api/src/main/java/com/idoceb00/elogap/repository/memory/InMemoryActivityRepository.java
package com.idoceb00.elogap.repository.memory;

import com.idoceb00.elogap.domain.Activity;
import com.idoceb00.elogap.domain.MatchResult;
import com.idoceb00.elogap.repository.ActivityRepository;
import org.springframework.stereotype.Repository;

import java.time.Duration;
import java.time.Instant;
import java.util.List;
import java.util.Optional;

@Repository
public class InMemoryActivityRepository implements ActivityRepository {

    private final List<Activity> data;

    public InMemoryActivityRepository() {
        Instant now = Instant.now();

        this.data = List.of(
                new Activity(
                        "match_1",
                        "player_1",
                        MatchResult.win,
                        "RANKED_SOLO",
                        "Ahri",
                        "MID",
                        10, 2, 8,
                        210,
                        1714,
                        24500,
                        32,
                        now.minus(Duration.ofHours(2))
                ),
                new Activity(
                        "match_2",
                        "player_1",
                        MatchResult.loss,
                        "RANKED_SOLO",
                        "Jinx",
                        "ADC",
                        4, 7, 6,
                        260,
                        1980,
                        30200,
                        18,
                        now.minus(Duration.ofHours(6))
                )
        );
    }

    @Override
    public List<Activity> list() {
        return data;
    }

    @Override
    public Optional<Activity> findById(String id) {
        return data.stream().filter(a -> a.id().equals(id)).findFirst();
    }
}